File: solutions/elixir/anagram/1/anagram.exs
defmodule Anagram do
  def match(word, list) do
    Enum.filter(list, fn(candidate_word) -> is_anagram(word, candidate_word) end)
  end

  def is_anagram(word, candidate_word) do
    has_same_letters(word, candidate_word) && !is_same_word(word, candidate_word)
  end

  def has_same_letters(word, candidate_word) do
    word_letters(word) == word_letters(candidate_word)
  end
  defp word_letters(word) do
    :lists.sort String.codepoints word |> String.downcase
  end

  def is_same_word(word, candidate_word) do
    String.downcase(word) == String.downcase(candidate_word)
  end
end

File: solutions/javascript/anagram/1/anagram.js
"use strict";
var Anagram = (function() {
  function Anagram(word) {
    var self = this;
    this.word = word;
    this.match = function(words) {
      var matches = words.reduce(function(collection, candidateWord) {
        if (testMatch(candidateWord)) { collection.push(candidateWord); }
        return collection;
      }, []);
      return matches;
    }
    function testMatch(candidateWord) {
      return reorderLetters(candidateWord) === reorderLetters(self.word);
    }
    function reorderLetters(word) {
      return word.split("").sort().join();
    }
  }
  return Anagram;
})();

module.exports = Anagram;